stacker/blueprints/variables/types.py

```diff
@@ -2,50 +2,100 @@
+
+        Complete resource definitions must be dictionaries, with the keys
+        identifying the resource titles, and the values being used as the
+        constructor parameters.
+
+        Parameter classes can be defined as dictionariy or a list of
+        dictionaries. In either case, the keys and values will be used directly
+        as constructor parameters.
+
         Args:
             defined_type (type): Troposphere type
-            many (Optional[bool]): Boolean indicating whether or not the type
-                accepts a single object or a list of objects.
+            many (Optional[bool]): Boolean indicating whether or not multiple
+                resources can be constructed. If the defined type is a
+                resource, multiple resources can be passed as a dictionary of
+                dictionaries
+                If it is a parameter class, multiple resources are passed as
+                a list.
+            optional (Optional[bool]): Boolean indicating if an undefined/null
+                configured value is acceptable, and can be passed to the
+                template instead of triggering an error.
 
         """
+
         self._validate_type(defined_type)
+        if many and optional:
+            raise ValueError("Cannot specify both `many` and `optional`")
+
         self._type = defined_type
         self._many = many
+        self._optional = optional
 
     def _validate_type(self, defined_type):
         if not hasattr(defined_type, "from_dict"):
             raise ValueError("Type must have `from_dict` attribute")
 
+    @property
+    def resource_name(self):
+        return (getattr(self._type, 'resource_name', None)
+                or self._type.__name__)
+
     def create(self, value):
         """Create the troposphere type from the value.
 
         Args:
-            value (dict): a dictionary of the resource name to a dictionary of
-                the values you want to pass to `from_dict` for the troposphere
+            value (Union[dict, list]): A dictionary or list of dictionaries
+                (see class documentation for details) to use as parameters to
+                create the Troposphere type instance.
+                Each dictionary will be passed to the `from_dict` method of the
                 type.
 
         Returns:
             Union[list, type]: Returns the value converted to the troposphere
                 type
 
         """
-        if not self._many and len(value) > 1:
-            message = (
-                "Only one resource can be provided for single "
-                "TroposphereType"
-            )
-            raise ValueError(message)
-
-        output = [self._type.from_dict(title, values) for title, values in
-                  value.items()]
-
-        return output if self._many else output[0]
+        if self._optional and not value:
+            return None
+
+        if hasattr(self._type, 'resource_type'):
```

Comment:
This seems to break some tests. Unfortunately it seems there are some inconsistencies in troposphere.